docs/audits/2024_q4/bdk_audit_report.md

```diff
@@ -0,0 +1,1031 @@
+crash bug which can be triggered by a third party, or a significant slowdown in performing some
+operations).
+
+A threat model i considered interesting for the purpose of this review was that of a server-side
+software using BDK to perform Bitcoin tasks and serving its clients through a public API. In this
+fairly standard scenario, a crash internal to the BDK library could lead to a Denial of Service for
+all users of the application.
+
+From my discussion with the developers of the library, remote Esplora and Electrum servers are
+trusted sources of information about the Bitcoin network. A rogue server providing invalid
+information is therefore considered out of scope. Likewise, i didn't focus on finding crashes that
+can be triggered by a malicious server (although i did report some).
+
+I have used test coverage reports to guide me throughout this audit. I join to this document a full
+report of the code coverage of all the unit tests from the library. It's available
+[here](https://darosior.github.io/bdk_coverage/unit_tests/coverage/home/darosior/projects/bdk/crates/index.html).
+
+![image](bdk_audit_test_coverage.png)
+
+We conducted this audit in three parts. First i carefully inspected the code of each module under
+review for logic errors, leveraging the unit test framework. I did not find it necessary to write
+and share new unit tests (besides some reproductions, see the details section) since the reviewed
+module already have extensive unit test coverage as can be observed from the above report. A second
+part of the audit was undertaken by @Pythcoiner (who works with us at Wizardsardine on Liana's QA)
+and consisted of manually testing the library, looking for surprising behaviour in corner cases.
+Finally, i wrote two fuzz targets (code is shared along with this report):
+- The first one exercised the `Wallet` interface under a number of conditions with data generated by
+  the fuzzer (apply an update, roundtrip to persistence, or create a transaction -then repeat). It's
+  more superfluous as it's general, but is a good way to assert some general invariants (for
+  instance "the balance is the same before and after round-tripping to persistence") and checking
+  nothing crashes if we just throw random data at the high-level interface.
+- The second one specifically targets the reorg logic in `bdk_chain` (the `merge_chains` function).
+  Based on the fuzzer's input it creates an initial chain and an updated chain, merges the two, then
+  asserts some invariants (it does connect when it should, etc).
+
+BDK is a large project and 10 days is not enough to perform a comprehensive review of the library.
+Only the 4 modules i focused on already amount to 15,777 lines of code (without accounting for the
+unit test directories). Here is a list of things i did not have time to do and think would be
+beneficial:
+- More fuzz coverage. I only started writing a couple fuzz targets after having reviewed the 15k
```

Comment:
Should we open a new issue considering the implementation of more fuzz targets?

Reply:
Yup, I opened this one: bitcoindevkit/bdk#1686